File: frontend/src/services/meetingMaster.ts
async function requestJson<T>(
  input: string,
  init: RequestInit = {},
): Promise<T> {
  const headers = new Headers(init.headers);

  if (!headers.has("Content-Type") && init.body) {
    headers.set("Content-Type", "application/json");
  }

  const response = await fetch(input, {
    ...init,
    headers,
    cache: "no-store",
  });

  if (!response.ok) {
    let message = `Request failed with status ${response.status}`;

    try {
      const data = await response.json();
      message = data?.detail || data?.message || message;
    } catch {
      // Keep default message.
    }

    throw new Error(message);
  }

  if (response.status === 204) {
    return undefined as T;
  }

  return (await response.json()) as T;
}

export type MeetingMaster = {
  meeting_id: number;
  meeting_name: string;
  meeting_type_id: number;
  meeting_type: string;
  client_id: number;
  client_code: string;
  audit_year: string;
  meeting_date: string;
  audit_start_date: string;
  audit_end_date: string;
  meeting_venue: string;
  meeting_note1: string;
  status: string;
  is_active: boolean;
  created_by?: string | null;
  updated_by?: string | null;
  created_at: string;
  updated_at: string;
};

export type MeetingMasterListResponse = {
  total: number;
  page: number;
  page_size: number;
  items: MeetingMaster[];
};

export type MeetingMasterPayload = {
  meeting_name: string;
  meeting_type_id: number;
  meeting_type: string;
  client_id: number;
  client_code: string;
  audit_year: string;
  meeting_date: string;
  audit_start_date: string;
  audit_end_date: string;
  meeting_venue: string;
  meeting_note1: string;
  status: string;
};

export type MeetingMasterUpdatePayload = Partial<MeetingMasterPayload> & {
  is_active?: boolean;
};

export type MeetingMasterListParams = {
  page?: number;
  page_size?: number;
  pageSize?: number;
  search?: string;
  sort_by?: string;
  sortBy?: string;
  sort_order?: "asc" | "desc";
  sortOrder?: "asc" | "desc";
  is_active?: boolean;
  isActive?: boolean;
};

function buildQuery(params: MeetingMasterListParams = {}): string {
  const normalizedParams = {
    page: params.page,
    page_size: params.page_size ?? params.pageSize,
    search: params.search,
    sort_by: params.sort_by ?? params.sortBy,
    sort_order: params.sort_order ?? params.sortOrder,
    is_active: params.is_active ?? params.isActive,
  };

  const searchParams = new URLSearchParams();

  Object.entries(normalizedParams).forEach(([key, value]) => {
    if (value === undefined || value === null || value === "") {
      return;
    }

    searchParams.set(key, String(value));
  });

  const query = searchParams.toString();
  return query ? `?${query}` : "";
}

export async function listMeetingMaster(
  params: MeetingMasterListParams = {},
): Promise<MeetingMasterListResponse> {
  return requestJson<MeetingMasterListResponse>(
    `/api/backend/meeting-master${buildQuery(params)}`,
  );
}

export async function createMeetingMaster(
  payload: MeetingMasterPayload,
): Promise<{ message: string; data: MeetingMaster | null }> {
  return requestJson<{ message: string; data: MeetingMaster | null }>(
    "/api/backend/meeting-master",
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
  );
}

export async function updateMeetingMaster(
  meetingId: number,
  payload: MeetingMasterUpdatePayload,
): Promise<{ message: string; data: MeetingMaster | null }> {
  return requestJson<{ message: string; data: MeetingMaster | null }>(
    `/api/backend/meeting-master/${meetingId}`,
    {
      method: "PATCH",
      body: JSON.stringify(payload),
    },
  );
}

export async function deactivateMeetingMaster(
  meetingId: number,
): Promise<{ message: string; data: MeetingMaster | null }> {
  return requestJson<{ message: string; data: MeetingMaster | null }>(
    `/api/backend/meeting-master/${meetingId}`,
    {
      method: "DELETE",
    },
  );
}

export async function restoreMeetingMaster(
  meetingId: number,
): Promise<{ message: string; data: MeetingMaster | null }> {
  return requestJson<{ message: string; data: MeetingMaster | null }>(
    `/api/backend/meeting-master/${meetingId}/restore`,
    {
      method: "PATCH",
    },
  );
}

export async function permanentDeleteMeetingMaster(
  meetingId: number,
): Promise<{ message: string; data: MeetingMaster | null }> {
  return requestJson<{ message: string; data: MeetingMaster | null }>(
    `/api/backend/meeting-master/${meetingId}/permanent`,
    {
      method: "DELETE",
    },
  );
}

File: backend/app/api/v1/meeting_master.py
from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.permissions import require_permission
from app.db.session import get_db
from app.models.user import User
from app.schemas.meeting_master import (
    MeetingMasterCreate,
    MeetingMasterListResponse,
    MeetingMasterMessageResponse,
    MeetingMasterResponse,
    MeetingMasterUpdate,
)
from app.services.meeting_master.meeting_master_service import MeetingMasterService


router = APIRouter(prefix="/meeting-master", tags=["Meeting Master"])


@router.get("", response_model=MeetingMasterListResponse)
async def list_meeting_master(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    search: str | None = None,
    is_active: bool | None = None,
    meeting_type: str | None = None,
    status_filter: str | None = Query(default=None, alias="status"),
    sort_by: str = "meeting_id",
    sort_order: str = "asc",
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("menu.meeting_master.view")),
):
    service = MeetingMasterService(db)

    return await service.list_meeting_master(
        page=page,
        page_size=page_size,
        search=search,
        is_active=is_active,
        meeting_type=meeting_type,
        status_filter=status_filter,
        sort_by=sort_by,
        sort_order=sort_order,
    )


@router.get("/{meeting_id}", response_model=MeetingMasterResponse)
async def get_meeting_master(
    meeting_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("menu.meeting_master.view")),
):
    service = MeetingMasterService(db)
    return await service.get_meeting_master(meeting_id)


@router.post(
    "",
    response_model=MeetingMasterMessageResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_meeting_master(
    payload: MeetingMasterCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("api.meeting_master.create")),
):
    service = MeetingMasterService(db)

    return await service.create_meeting_master(
        payload=payload,
        created_by=current_user.user_id,
    )


@router.patch("/{meeting_id}", response_model=MeetingMasterMessageResponse)
async def update_meeting_master(
    meeting_id: int,
    payload: MeetingMasterUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("api.meeting_master.update")),
):
    service = MeetingMasterService(db)

    return await service.update_meeting_master(
        meeting_id=meeting_id,
        payload=payload,
        updated_by=current_user.user_id,
    )


@router.delete("/{meeting_id}", response_model=MeetingMasterMessageResponse)
async def delete_meeting_master(
    meeting_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("api.meeting_master.delete")),
):
    service = MeetingMasterService(db)

    return await service.deactivate_meeting_master(
        meeting_id=meeting_id,
        updated_by=current_user.user_id,
    )


@router.patch("/{meeting_id}/restore", response_model=MeetingMasterMessageResponse)
async def restore_meeting_master(
    meeting_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("api.meeting_master.restore")),
):
    service = MeetingMasterService(db)

    return await service.restore_meeting_master(
        meeting_id=meeting_id,
        updated_by=current_user.user_id,
    )


@router.delete(
    "/{meeting_id}/permanent",
    response_model=MeetingMasterMessageResponse,
)
async def permanent_delete_meeting_master(
    meeting_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(
        require_permission("api.meeting_master.permanent_delete")
    ),
):
    service = MeetingMasterService(db)
    return await service.permanent_delete_meeting_master(meeting_id)
